src/main/java/net/neoforged/neoforge/transfer/handlers/wrappers/items/ItemContainerContentsResourceHandler.java

```diff
@@ -0,0 +1,169 @@
+    }
+
+    @Override
+    public int getCapacity(int index, ItemResource resource) {
+        Objects.checkIndex(index, size());
+        if (resource.isEmpty()) return Item.ABSOLUTE_MAX_STACK_SIZE;
+        return Math.min(resource.getMaxStackSize(), Item.ABSOLUTE_MAX_STACK_SIZE);
+    }
+
+    @Override
+    public boolean isValid(int index, ItemResource resource) {
+        ItemResource current = getResource(index);
+        return current.isEmpty() || current.equals(resource);
+    }
+
+    @Override
+    public boolean supportsInsertion(int index) {
+        return true;
+    }
+
+    @Override
+    public boolean supportsExtraction(int index) {
+        return true;
+    }
+
+    //TODO the resource should be in theory ignored here, but I want to be sure.
+    public void set(int index, ItemResource resource, int amount) {
+        ItemContainerContents contents = getContents();
+        contents.getStackInSlot(index).setCount(amount);
+        try (Transaction tx = TransactionManager.open(null)) {
+            setAndValidate(contents, amount, tx);
+            tx.commit();
+        }
+    }
+
+    @Override
+    public int insert(int index, ItemResource resource, int amount, TransactionContext context) {
+        if (ResourceHandlerUtil.isEmpty(resource, amount) || !isValid(index, resource)) return 0;
+        ItemContainerContents contents = getContents();
+        ItemStack stack = contents.getStackInSlot(index);
+        if (stack.isEmpty()) {
+            amount = Math.min(amount, resource.getMaxStackSize());
+
+            return setAndValidate(contents.with(index, resource, amount), amount, context);
+        } else if (resource.is(stack) && stack.getCount() < resource.getMaxStackSize()) {
+            int newAmount = Math.min(stack.getCount() + amount, resource.getMaxStackSize());
+            amount = newAmount - stack.getCount();
+//            if (action.isExecuting())
+            stack.grow(amount);
+            return setAndValidate(contents, amount, context);
+        }
+        return 0;
+    }
+
+    @Override
+    public int insert(ItemResource resource, int amount, TransactionContext context) {
+        if (ResourceHandlerUtil.isEmpty(resource, amount)) return 0;
+        ItemContainerContents contents = getContents();
+        int remaining = amount;
+        for (int i = 0; i < size; i++) {
+            ResourceStack<ItemResource> stack = contents.getStackInSlot(i).immutable();
+            if (stack.isEmpty() || !stack.resource().equals(resource) || stack.amount() >= resource.getMaxStackSize()) continue;
+            int toInsert = Math.min(remaining, resource.getMaxStackSize() - stack.amount());
+            contents = contents.with(i, resource, stack.amount() + toInsert);
+            remaining -= toInsert;
+        }
+        for (int i = 0; i < size; i++) {
+            ResourceStack<ItemResource> stack = contents.getStackInSlot(i).immutable();
+            if (!stack.isEmpty()) continue;
```

Comment:
Just add a helper to `ItemContainerContents` next to the `getStackInSlot` for `isSlotEmpty(index)`, and use that instead of copying the stack and then turning it into an immutable resource stack just to check if it is empty...